### ISSUE TITLE
Figure out what broke the JS builds

### ISSUE THREAD

**Issue by tamasfe** [tamasfe/taplo#175]

**Comment by buzztaiki** [comment]
I think this error is due to the wasm build using an old toolchain. 

https://github.com/tamasfe/taplo/blob/5a4900b9862b7f18202b8b38b120cf5a0549b474/.github/workflows/ci.yaml#L533

When I built in the my local environment with latest stable toolchain, this error disappeared, but I got the following new error:

<details>

```console
 $  (cd js/lsp/ && yarn build)
yarn run v1.22.17
$ yarn rollup --silent -c rollup.config.js
$ /home/taiki/dev/src/github.com/tamasfe/taplo/js/lsp/node_modules/.bin/rollup --silent -c rollup.config.js
   Compiling taplo-lsp v0.2.1 (/home/taiki/dev/src/github.com/tamasfe/taplo/taplo-lsp)
error: future cannot be sent between threads safely
   --> taplo-lsp/src/lib.rs:357:10
    |
357 |         .on_request::<request::Completion, _>(handlers::completion)
    |          ^^^^^^^^^^ future returned by `completion` is not `Send`
    |
    = help: within `impl futures::Future`, the trait `std::marker::Send` is not implemented for `*mut u8`
note: future is not `Send` as this value is used across an await
   --> taplo-lsp/src/lib.rs:248:21
    |
245 |             let res = client.get(path).send().await?;
    |                 --- has type `reqwest::Response` which is not `Send`
...
248 |             let w = context.world().lock().await;
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ await occurs here, with `res` maybe used later
...
285 |         } else if path.starts_with("file://") {
    |         - `res` is later dropped here

error: future cannot be sent between threads safely
   --> taplo-lsp/src/lib.rs:357:10
    |
357 |         .on_request::<request::Completion, _>(handlers::completion)
    |          ^^^^^^^^^^ future returned by `completion` is not `Send`
    |
    = help: within `impl futures::Future`, the trait `std::marker::Send` is not implemented for `Rc<RefCell<wasm_bindgen_futures::Inner>>`
note: future is not `Send` as it awaits another future which is not `Send`
   --> taplo-lsp/src/external/wasm32/mod.rs:40:1
    |
40  | #[wasm_bindgen]
    | ^^^^^^^^^^^^^^^ await occurs here on type `JsFuture`, which is not `Send`
    = note: this error originates in the attribute macro `wasm_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)

error: future cannot be sent between threads safely
   --> taplo-lsp/src/lib.rs:358:10
    |
358 |         .on_request::<request::HoverRequest, _>(handlers::hover)
    |          ^^^^^^^^^^ future returned by `hover` is not `Send`
    |
    = help: within `impl futures::Future`, the trait `std::marker::Send` is not implemented for `*mut u8`
note: future is not `Send` as this value is used across an await
   --> taplo-lsp/src/lib.rs:248:21
    |
245 |             let res = client.get(path).send().await?;
    |                 --- has type `reqwest::Response` which is not `Send`
...
248 |             let w = context.world().lock().await;
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ await occurs here, with `res` maybe used later
...
285 |         } else if path.starts_with("file://") {
    |         - `res` is later dropped here

error: future cannot be sent between threads safely
   --> taplo-lsp/src/lib.rs:358:10
    |
358 |         .on_request::<request::HoverRequest, _>(handlers::hover)
    |          ^^^^^^^^^^ future returned by `hover` is not `Send`
    |
    = help: within `impl futures::Future`, the trait `std::marker::Send` is not implemented for `Rc<RefCell<wasm_bindgen_futures::Inner>>`
note: future is not `Send` as it awaits another future which is not `Send`
   --> taplo-lsp/src/external/wasm32/mod.rs:40:1
    |
40  | #[wasm_bindgen]
    | ^^^^^^^^^^^^^^^ await occurs here on type `JsFuture`, which is not `Send`
    = note: this error originates in the attribute macro `wasm_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)

error: future cannot be sent between threads safely
   --> taplo-lsp/src/lib.rs:359:10
    |
359 |         .on_request::<request::DocumentLinkRequest, _>(handlers::links)
    |          ^^^^^^^^^^ future returned by `links` is not `Send`
    |
    = help: within `impl futures::Future`, the trait `std::marker::Send` is not implemented for `*mut u8`
note: future is not `Send` as this value is used across an await
   --> taplo-lsp/src/lib.rs:248:21
    |
245 |             let res = client.get(path).send().await?;
    |                 --- has type `reqwest::Response` which is not `Send`
...
248 |             let w = context.world().lock().await;
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ await occurs here, with `res` maybe used later
...
285 |         } else if path.starts_with("file://") {
    |         - `res` is later dropped here

error: future cannot be sent between threads safely
   --> taplo-lsp/src/lib.rs:359:10
    |
359 |         .on_request::<request::DocumentLinkRequest, _>(handlers::links)
    |          ^^^^^^^^^^ future returned by `links` is not `Send`
    |
    = help: within `impl futures::Future`, the trait `std::marker::Send` is not implemented for `Rc<RefCell<wasm_bindgen_futures::Inner>>`
note: future is not `Send` as it awaits another future which is not `Send`
   --> taplo-lsp/src/external/wasm32/mod.rs:40:1
    |
40  | #[wasm_bindgen]
    | ^^^^^^^^^^^^^^^ await occurs here on type `JsFuture`, which is not `Send`
    = note: this error originates in the attribute macro `wasm_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `taplo-lsp` due to 6 previous errors
Error: Compiling your crate to WebAssembly failed
Caused by: failed to execute `cargo build`: exited with exit code: 101
  full command: "cargo" "build" "--lib" "--target" "wasm32-unknown-unknown"
[!] (plugin rust) Error: Rust compilation failed
../../taplo-lsp/Cargo.toml
Error: Rust compilation failed
    at wasm_pack (/home/taiki/dev/src/github.com/tamasfe/taplo/js/lsp/node_modules/@wasm-tool/rollup-plugin-rust/index.js:171:19)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Promise.all (index 0)
    at build (/home/taiki/dev/src/github.com/tamasfe/taplo/js/lsp/node_modules/@wasm-tool/rollup-plugin-rust/index.js:339:22)
    at ModuleLoader.addModuleSource (/home/taiki/dev/src/github.com/tamasfe/taplo/js/lsp/node_modules/rollup/dist/shared/rollup.js:18279:30)
    at ModuleLoader.fetchModule (/home/taiki/dev/src/github.com/tamasfe/taplo/js/lsp/node_modules/rollup/dist/shared/rollup.js:18335:9)
    at async Promise.all (index 0)
    at ModuleLoader.fetchStaticDependencies (/home/taiki/dev/src/github.com/tamasfe/taplo/js/lsp/node_modules/rollup/dist/shared/rollup.js:18360:34)
    at async Promise.all (index 0)
    at ModuleLoader.fetchModule (/home/taiki/dev/src/github.com/tamasfe/taplo/js/lsp/node_modules/rollup/dist/shared/rollup.js:18337:9)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
</details>

I think it's because the wasm code of lsp-async-stub was removed in https://github.com/tamasfe/taplo/commit/adfc881b4d66ae98f8496a06710386b78386f54b.

**Comment by tamasfe** [comment]
Ah yes, thanks. There were 2 issues, one of which was the outdated toolchain did not support the new resolver and rust 2021 edition. The other was that while refactoring `lsp-async-stub`, I forgot that a lot of types are `!Send` in WASM context. Both of these should be fixed in 7afeb39df0ba328f148606b0b0b8070e56e87a82.

**Comment by tamasfe** [comment]
This seems to be more complicated... Now there are several cases where futures are still `!Send` but are required for a multi-threaded environment for native tokio builds, I have no idea how they worked until this point though. At least the wasm builds work now.
